### DISCUSSION TITLE
hide crossing

### DISCUSSION THREAD

**Discussion by kianzarrin** [CitiesSkylinesMods/TMPE#1503] [General]
@aubergine10 When I wanted to merge hide crossing into TMPE other team members had lost interest.
are you still interested?

**Comment by originalfoo** [comment]
@kianzarrin Yes, it would be great addition IMO.

It's weird that we need external mod to do it, given that TMPE has the most commonly used UI for toggling crossings on/off.

**Reply by kianzarrin** [reply]
Currently hide crossing mod is the old way to do it. Hide crossing does not work for lods and it does not work for every road. The new way to do it is for the asset editor to use AN. But that requires asset creators to use AN and to update their roads which I don't think is going to happen.

Two other mods that should go inside TMPE or AN is Hide unconnected tracks and Direct connect roads.

@chameleon-tbn had some ideas about how to do these things.

**Reply by krzychu124** [reply]
Like I said some time ago, I would like to see it a module (new dll, we can enable internal visibility). It will be easier to maintain and also integrate. Question remains: how to handle other mods that require Hide Crossing (NCR?) and how to handle the case when user already have standalone version of Hide crossing mod (subscribed or local copy)

**Reply by chameleon-tbn** [reply]
@krzychu124 i absolutely agree, that all three should be Part of TMPE.

Disable Crosswalk should hide crossings, and in options it could be possible to turn off this behaviour (but imho even this would Not be needed)

Same for Direct Connect Road and Hide unconnected Tracks as they already Setup with the UI of TMPE.

I would simple do a Pop-up at Main Menu load that tells people, the old mods are now incompatible and they need to unsub them, as functions are now build in. That's the way Quistar do with his mods and imho only a link to WS Page should be added (and in Case of local Mod Point there). As Long as they are still there, TMPE could either deactivate them, deactivate the build in function, or disable itself - but imho this is Not needed. Imho the separate Pop-up IS enough.

**Reply by kianzarrin** [reply]
Where do I put the settings for each sub-module? There doesn't seem to be space for more tabs :).

**Reply by originalfoo** [reply]
> and in options it could be possible to turn off this behaviour (but imho even this would Not be needed)

It uses extra RAM so IMO user should be able to turn it off; there's lots of users who are barely scraping by with limited RAM.

> Where do I put the settings for each sub-module? There doesn't seem to be space for more tabs :).

How many settings are needed - is it just one checkbox per module, or are there additional settings that need to be added?

**Reply by kianzarrin** [reply]
Additional settings about exclusions, mesh generation, ...

**Reply by kianzarrin** [reply]
> It uses extra RAM

Continues junction barrier mod (or should I call it hide broken medians) has 2 functions:
1- hide broken medians when road does have DC.  
2- mesh generations where road does not have DC.
3- it also interacts with AN to know which DC to ignore.

The first one does not use RAM and might be useful in #1223

**Reply by kianzarrin** [reply]
so maybe some part of the code could be inside TMPE and some other part of the code could be as a sub-module?

**Reply by krzychu124** [reply]
Hmm for settings... 
Maybe instead of new tabs we should try to build internal options screen.
I see it like we have a special tab which contains list of buttons (with label and icon, layout is grid-like) and you can click on specific button (say Hide Crossing, with icon) to see related settings and in the upper corner of that whole screen display either "<= back" button or just and arrow "<=" to return to main view (selection of buttons). So when you click on any button we just hide everything (current tabs view) and show container with new settings specific for selected option on the first screen. Theoretically we could reuse existing functions like `OnSettingsUI`, it expects `UIHelper` so we could create `UIHelper` from new panel. Limits required changes to minimum, rest could be just wrapped nicely.  IUserMod should be swapped to TMPE version so we could reuse as much as possible without making too many changes. Extension classes like `SerializationExtention/LoadingExtension` will be handled by the game with ease (I'm pretty sure it doesn't care how many implementation you've created in your mod so it'll take everything).

Ugly example 🙃  (HC - hide crossing, DC - Direct Connect, either abbrev. or full name, doesn't matter):
![image](https://user-images.githubusercontent.com/19638970/161543851-3ca33b38-fb5a-469a-88ab-550b751da113.png)

**Reply by originalfoo** [reply]
> Probably need to do this scale of change in 11.6.6.x branch; 11.6.5.x contains lots of bug fixes and improvements we need to get to STABLE

Ditch tabs completely. Could have a settings "home" page something like (screenshot from cslmodding.info website):

![image](https://user-images.githubusercontent.com/1386719/161563105-be637806-6529-4705-a35a-5ca792a5ac39.png)

EDIT: Layout would probably need to be more compact, eg. text at side of icon. Screen space is limited.

Each panel representing an options "group" (we've already got most stuff split in to group files, but can easily move stuff around, especially once the option UI definitions get moved to `Options.cs`. So the option group class will define title/tooltip/icon of its panel on the "home" page.

EDIT: Bonus round - do the group `AddUI()` on-demand. Faster loading times due to not instantiating UI components during load.

Once in a group, we could have a nav bar at the top - something like you see in web browsers (hasty mockup made by customising my browser toolbar lol):

![image](https://user-images.githubusercontent.com/1386719/161565179-68f0e099-1aa7-44cb-8257-ff69d3ffd5c8.png)

That allows user to return to the main "home" page (panel grid), scroll through the groups (left/right arrows), save as global default, load from global default, and reload main config. At some point in future "save as..." - allowing user to create custom/sharable options profiles. We could have buttons on there eg. for "what's new", "user guide", "language", etc?

EDIT: The nav bar at top could exist on the "Home" screen too, but with options such as (very rough mockup):

![image](https://user-images.githubusercontent.com/1386719/161577993-2f212200-41c9-4a64-a83e-5758cd6e0b4b.png)

**Reply by kianzarrin** [reply]
also possible to have multirow tabs like windows

**Reply by krzychu124** [reply]
Or... scrollable tabs, would be better than multirow. 
I would like to remove tabs since sooner or later we run out of space and variant with clickable tiles might help us dividing setting into _simplified_ and _advanced/all_ options screen

> Probably need to do this scale of change in 11.6.6.x branch; 11.6.5.x contains lots of bug fixes and improvements we need to get to STABLE

Yes but 11.6.5.x also introduced breaking change so users from TEST won't be able to load savegames using STABLE version once we release it to 11.6.5.2 to the TEST...

Integrating HC/DC etc. should not be done before releasing 11.6.5.x to Stable.

**Reply by originalfoo** [reply]
tagging https://github.com/CitiesSkylinesMods/TMPE/issues/652 which is related

**Comment by kianzarrin** [comment]
~~m_paths is set to null one asset is being set even though it is not null before save.
I put a null guard.
I don't save if there is nothing to save~~

**Reply by originalfoo** [reply]
Misplaced comment?

**Reply by krzychu124** [reply]
> Misplaced comment?

yeah, sounds like a comment to that bug related with Asset editor
